scalyr_agent/log_processing.py

```diff
@@ -493,8 +547,22 @@ def readline(self, current_time=None):
         #        assert expected_size == actual_size, ('Mismatch between expected and actual size %ld %ld',
         #                                              expected_size, actual_size)
 
-        # check to see if we need to parse the line as json
-        if self.__parse_as_json:
+        # check to see if we need to parse the line as cri.
+        # Note: we don't handle multi-line lines.
+        if self.__parse_format == 'cri':
+            timestamp, stream, tags, message = _parse_cri_log( result.line )
+            if message is None:
+                log.warn("Didn't find a valid log line in CRI format for log %s.  Logging full line." %
```

Comment:
I created a test cluster in EKS and deployed your branch.   This is docker (not containerD).  The only environment variables I've set are May have found a bug? =>  `kubernetes_monitor.log` is empty.  `agent.log` suggests it has gone into CRI mode (but this is docker):

```
2019-04-19 01:11:28.672Z ERROR [monitor:kubernetes_monitor] [kubernetes_monitor.py:768] Error connecting to kubelet API endpoint - HTTPConnectionPool(host='192.168.45.216', port=10255): Max retries exceeded with url: /pods (Caused by NewConnectionError('<scalyr_agent.third_party.requests.packages.urllib3.connection.HTTPConnection object at 0x7fca8dc52fd0>: Failed to establish a new connection: [Errno 111] Connection refused',)). The Scalyr Agent will now monitor containers from the filesystem
...
2019-04-19 01:11:38.182Z WARNING [core] [log_processing.py:556] Didn't find a valid log line in CRI format for log /var/log/containers/currencyservice-668f49f985-kwlnm_default_server-a50fbbd13977584b58efa8f96a6be4c48784edb1029c41394036446aa1fd8d08.log.  Logging full line.
...
```

![image](https://user-images.githubusercontent.com/48775713/56399449-d84ea480-6202-11e9-8245-bdcf8343a375.png)

I also pulled in my status-v PR which verifies only the following env vars set:

![image](https://user-images.githubusercontent.com/48775713/56400012-5b252e80-6206-11e9-9925-ae74efcd48dd.png)

Reply:
Can you invite me to look at the logs on your account?

Reply:
I've emailed you a .tgz of configs, logs, statusv output

Reply:
Update:  I deployed the exact image on a different test k8s docker cluster (that @czerwingithub normally uses). This time it works fine.  My scripts are pretty much exactly the same (for both clusters).  The only difference might possibly be the RBAC service account settings which I have not changed for a long time in the failing cluster? (I will inspect them for differences).

Regardless, it sounds to me what might be happening is agent fails to ping kublet API, then causing it  to enter CRI parsing mode.  The files in `/var/log/containers` are there but are in JSON format hence the errors.

Reply:
> it sounds to me what might be happening is agent fails to ping kublet API, then causing it to enter CRI parsing mode

I'm not sure that's it, because even if it enters CRI mode, docker provides CRI support and the monitor should detect that it is docker under CRI and still parse as json.  The `parse_format` is distinct from whether or not we are checking CRI locations for container and log information.

The only reason it should be trying to parse the logs in CRI is if it didn't detect that the containers were running under docker, so I'd like to try and figure out what it thought the container runtime was.

Reply:
Aaaand the container runtime is of type `unknown` :-/

Which is not `docker` and so it defaults to parsing in CRI format.

Let's move this to a discussion more amenable to back-and-forth conversation.  If you're still up, I'll ping you in Slack.

Reply:
The agent now checks to see if it can detect docker by other means, and also sets the parse format to json if that's the case.

Reply:
I deployed a fresh image with your latest branch to my test EKS cluster.  Verified JSON docker json logs are being parsed as json.

![image](https://user-images.githubusercontent.com/48775713/56515544-f95a1280-64ec-11e9-9264-518de3270f63.png)

Note to self:  When we have automated end-to-end testing for k8s, the test should verify JSON is parsed correctly.